Source/WebCore/rendering/style/StyleFilterImage.cpp

```diff
@@ -165,4 +165,16 @@ void StyleFilterImage::imageChanged(CachedImage*, const IntRect*)
         client->imageChanged(static_cast<WrappedImagePtr>(this));
 }
 
+void StyleFilterImage::didAddClient(RenderElement& renderer)
```

Comment:
I am a little worried about what this doing when m_image is a StyleCachedImage.

If you look in the header, StyleFilterImage.h, you will see a FIXME explaining some of the current issues these client relationships (and the inherited weird use of CachedImage).

To back up, the basic purpose of these clients is to allow "parent" images to know when a "child" image has changed, so that it can invalidate itself, and for the owner of the image "tree" (usually something in the render tree, but that should be a requirement we get rid of probably, as there are places like HTML canvas that want to use them), to know when anything in the image tree changes so it can invalidate itself.

When there was only StyleCachedImage, it made sense to just reuse `CachedImageClient` for this. That even made sense for something like `filter(src(image.png), invert(1))` since the only thing the ultimate owner cared about was making sure to invalidate when `src(image.png)` loaded.

The design I was thinking about for this was to add a new StyleImageClient interface that could be used both internally in the StyleImage "tree", and by the render tree. 

It is totally possible that would be overkill, but I think once we start talking about things like StyleCrossfadeImage, where a StyleImage can have two StyleImage children, trying to use CachedImage doesn't really work anymore.